sdk_v2/kubeflow/training/api/training_client.py

```diff
@@ -12,7 +12,553 @@
+                    for job in item["spec"]["template"]["spec"]["replicatedJobs"]:
+                        if job["name"] == constants.JOB_TRAINER_NODE:
+                            pod_spec = job["template"]["spec"]["template"]["spec"]
+                            for container in pod_spec["containers"]:
+                                if container["name"] == constants.CONTAINER_TRAINER:
+                                    if "resources" in container:
+                                        resources = client.V1ResourceRequirements(
+                                            **container["resources"]
+                                        )
+
+                    # TODO (andreyvelich): Currently, only Torch is supported for NumProcPerNode.
+                    num_procs = None
+                    if "torch" in item["spec"]["mlPolicy"]:
+                        num_procs = item["spec"]["mlPolicy"]["torch"]["numProcPerNode"]
+
+                    # Get the device count per Trainer node.
+                    # TODO (andreyvelich): Currently, we get the device type from
+                    # the runtime labels.
+                    _, device_count = utils.get_container_devices(resources, num_procs)
+                    if device_count != constants.UNKNOWN:
+                        device_count = str(
+                            int(device_count)
+                            * int(item["spec"]["mlPolicy"]["numNodes"])
+                        )
+
+                    runtime = types.Runtime(
+                        name=item["metadata"]["name"],
+                        phase=(
+                            item["metadata"]["labels"][constants.PHASE_KEY]
+                            if constants.PHASE_KEY in item["metadata"]["labels"]
+                            else constants.UNKNOWN
+                        ),
+                        device=(
+                            item["metadata"]["labels"][constants.DEVICE_KEY]
+                            if constants.DEVICE_KEY in item["metadata"]["labels"]
+                            else constants.UNKNOWN
+                        ),
+                        device_count=device_count,
+                    )
+
+                    result.append(runtime)
+        except multiprocessing.TimeoutError:
+            raise TimeoutError(
+                f"Timeout to list {constants.CLUSTER_TRAINING_RUNTIME_KIND}s "
+                f"in namespace: {self.namespace}"
+            )
+        except Exception:
+            raise RuntimeError(
+                f"Failed to list {constants.CLUSTER_TRAINING_RUNTIME_KIND}s "
+                f"in namespace: {self.namespace}"
+            )
+
+        return result
+
+    def train(
+        self,
+        runtime_ref: str,
+        trainer: Optional[types.Trainer] = None,
+        dataset_config: Optional[types.HuggingFaceDatasetConfig] = None,
+        model_config: Optional[types.HuggingFaceModelInputConfig] = None,
+    ) -> str:
```

Comment:
I refactored this API based on @akshaychitneni's suggestion above.
So the current submission looks as follows:
Pre-training:
```python

# Create the TrainJob with two PyTorch nodes. Every node has two GPUs.
job_id = TrainingClient().train(
    trainer=Trainer(
        func=train_pytorch_model,
        func_args={"lr": "0.01"},
        num_nodes=3,
        resources_per_node={"gpu": 2}
    ),
    runtime_ref=torch_runtime,
)
```
Fine-Tuning:
```python
from kubeflow.training import HuggingFaceDatasetConfig, LoraConfig, Trainer, FineTuningConfig

job_id = TrainingClient().train(
    dataset_config=HuggingFaceDatasetConfig(
        storage_uri="tatsu-lab/alpaca",
    ),
    trainer=Trainer(
        fine_tuning_config=FineTuningConfig(
            lora_config=LoraConfig(r=4),
        ),
       num_nodes=5,
    ),
    runtime_ref=llm_runtime,
)
```

What do you think about it @tenzen-y @droctothorpe @Electronic-Waste @seanlaii @deepanker13 @tenzen-y ?

Reply:
Another solution could be to separate `train()` and `fine_tune()` APIs, so `train()` can accept Training Function + Nodes Scaling, `fine_tune()` can accept the Fine-Tuning configuration + Nodes Scaling, Model Config, and Dataset Config.

```python
job_id = TrainingClient().train(
    func=train_pytorch_model,
    func_args={"lr": "0.01"},
    num_nodes=3,
    resources_per_node={"gpu": 2}
    runtime_ref=torch_runtime,
)

job_id = TrainingClient().fine_tune(
    dataset_config=HuggingFaceDatasetConfig(
        storage_uri="tatsu-lab/alpaca",
    ),
    fine_tuning_config=FineTuningConfig(
        lora_config=LoraConfig(r=4),
        num_nodes=5,
    ),
    runtime_ref=llm_runtime,
)
```
Any thoughts ?

Reply:
From my perspective, the former approach is better because:

1. We define all these configs like `num_nodes` in `Trainer` in `/pkg/api/v2alph1/trainjobs_types.go`. It will be more consistent and intuitive if we adopt the former approach.
2. A unified interface would be easier to understood and used by users.

And just a thoughts, we could define a unified PEFT config field like those in [llama_recipes](https://github.com/meta-llama/llama-recipes/blob/2f72bcec0d0c276e1c7304441a537eec3baf17f6/src/llama_recipes/utils/config_utils.py#L43-L45):

```python
job_id = TrainingClient().train(
    dataset_config=HuggingFaceDatasetConfig(
        storage_uri="tatsu-lab/alpaca",
    ),
    trainer=Trainer(
        fine_tuning_config=FineTuningConfig(
            peft_config=LoraConfig(r=4), # Optional[Union[LoraConfig, AdapterPromptConfig, PrefixTuningConfig]]
            sharding_config=FsdpConfig(...) # Optional[Union[FsdpConfig, ZeroConfig]]
        ),
        num_nodes=5,
    ),
    runtime_ref=llm_runtime,
)
```

Reply:
Yeah, I think it is a good idea @Electronic-Waste , since we can support qLoRA and other configs under PEFT config.

Reply:
SGTM